#160618097 Add JWT sign-up and sign-in functionality

AdinoyiSadiq opened the pull request:
### What does this PR do?
This pull request adds the sign-up and sign-in functionality for users using JWT

#### Description of Task to be completed?
Have the following endpoints working
- `/api/auth/signup`
- `/api/auth/signin`

Return a token on successful registration or authentication
Create an authenticated test route

#### How should this be manually tested?
- Setup the repo
- run `npm install`
- run `npm run test`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#160618097](https://www.pivotaltracker.com/story/show/160618097)

#### Screenshots (if appropriate)
N/A

#### Questions:
None

Commit:
feat(test): add initial signup test

- add an initial failing test for the user signup route

[Delivers #160618097]

Commit:
feat(signup): add user jwt signup functionality

- enable a user to signup to the application using firstname, lastname, email and password
- the user should receive a token on successful registration

[Delivers #160618097]

Commit:
feat(test): add passing signup test

[Delivers #160618097]

Commit:
feat(test): add failing tests for signup form validation

[Delivers #160618097]

Commit:
feat(validate): add signup form validation

- add validation scripts to test signup form for incorrectly filled fields

[Delivers #160618097]

AdinoyiSadiq committed:
feat(test): add initial signin failing test

[Delivers #160618097]

AdinoyiSadiq committed:
feat(signin): add signin functionality for users

- enable users to signin with email and password and get a token after authentication

[Delivers #160618097]

AdinoyiSadiq committed:
feat(test): add failing signin test

[Delivers #160618097]

AdinoyiSadiq committed:
feat(validate): add signin form validation

- add signin form validation to ensure that the user enters the correct values

[Delivers #160618097]

AdinoyiSadiq committed:
feat(test): add failing test for autheticated route

[Delivers #160618097]

AdinoyiSadiq committed:
feat(auth): add authenticated route

- add middleware to restrict access to certain routes to only authenticated users

[Delivers #160618097]

AdinoyiSadiq committed:
feat(merge): merge staging onto this branch

[Delivers #160618097]
Merge branch 'staging' into feature/160618097/jwt-login

damafeez commented:
You can consider fixing linting errors to allow HoundCI pass.
You should also add `"postinstall": "./node_modules/.bin/sequelize db:migrate"` script in `package.json` to enable heroku run migration before server start

AdinoyiSadiq committed:
feat(eslint): ensure that airbnb conventions are followed

[Delivers #160618097]

AdinoyiSadiq committed:
feat(feedback): implement feedback from Technical Team Lead

- refactor signup and signin controllers to follow single responsibility principle
- rename the requireAuth middleware to 'authenticate'
- rename the generateString function to 'generateErrorString' in the validate file
- refactor the validate function to follow the single responsibility principle
- create a user helper file to handle finding a user in the database

[Delivers #160618097]

ngozi-ekekwe commented on controllers/authController.js:
```diff
@@ -0,0 +1,55 @@
+
+    const encryptedPassword = await utility.encryptPassword(password);
+
+    User.create({
```
refactor this please using a spread operator

ngozi-ekekwe commented on middleware/authenticate.js:
```diff
@@ -0,0 +1,18 @@
+const jwt = require('jsonwebtoken');
```
rename this file to authentication
